pipeline/run_pipeline.py

```diff
@@ -227,41 +242,43 @@ def plotting_refusal_with_activations(
 
 
 def computing_dot_product(
-    data_list,
-    refusal_direction,
-    average_dot_products,
-    resultant_dot_products,
-    activations_dir,
-    layers,
-    harmful_train,
-):
-    data_list = [normalize_vector(data) for data in data_list]
-
+    model_base,
+    layer_wise_activations: List[torch.Tensor],
+    refusal_direction: torch.Tensor,
+    average_dot_products: List[torch.Tensor],
+    resultant_dot_products: List[torch.Tensor],
+    activations_dir: str, 
+    layers: int, 
+    harmful_train: List[torch.Tensor]
+) -> None:
+    layer_wise_activations = [normalize_vector(data,model_base) for data in layer_wise_activations]
+
+    #layer_wise_activations is list, converting it to tensors for matching shape
+    layer_wise_activations_tensor = torch.stack(layer_wise_activations)
+    assert layer_wise_activations_tensor.shape == (len(harmful_train), model_base.model.config.num_hidden_layers, model_base.model.config.hidden_size)
+        
+    device=refusal_direction.device
     """Calculate and save dot products, then generate and save a DataFrame with results."""
     for i in range(layers):
```

Comment:
There is no need to go through this for loop. You can use torch function to do everything without any for loop (which is way more efficient as torch functions are written in C++ and they operations in parallel across thousands of gpu-cores). With the below suggestions, you should get an idea on how to remove this for loop also.

Reply:
Should I remove all for loop then?

Reply:
using einsum?

Reply:
yo facing some issues, here
so to pass it on normalize function, i have to use for loop and the output of that is a list, and to convert the list back to tensor( because einsum accepts only tensor) I have to use torch.stack

So, either I introduce 
1)another variable to store and not use stack / 
2)use for loop and then stack the output

#Method 1
```
normalized_resultant_vectors = torch.zeros_like(resultant_vectors)
    for i in range(resultant_vectors.shape[0]):
        normalized_resultant_vectors[i] = normalize_vector(resultant_vectors[i], model_base)
        
```
#Method 2
```
normalized_resultant_vectors = [normalize_vector(resultant_vectors[i], model_base) for i in range(resultant_vectors.shape[0])]
normalized_resultant_vectors_tensor = torch.stack(normalized_resultant_vectors)
```

Reply:
Einsum is great if you know how to use it. However, you can do most things directly with torch and not require einsum (unless you're doing complicated stuff that torch doesn't provide functions for which is very rare). Just go through some tutorial on how to use pytorch.

```
normalized_resultant_vectors = normalize_vector(resultant_vectors, model_base)
```

You're adding for loops unnecessarily. If you're using a for loop and stack afterwards, it most likely means you're doing it wastefully and you can directly use torch function to do what you want. In this case, `normalize_vector` anyway normalizes along the last dim, so why go through a for loop and stack again when you can do it all in one torch operation?